### PR TITLE
Add JS library for interacting with fields inside a CRUD form

### PR DESCRIPTION
**Pull request by tabacitu**
### BEFORE - What was wrong? What was happening before this PR?

When you wanted to show/hide/toggle/change fields with JS, you had to write quite a bit of JS / jQuery code. For each field type... you'd have to think about the proper selector, try things out. Then when you manipulate it, you don't know where the info is stored (is it inside the main input/select/textarea? inside a hidden input? etc) and you don't know WHAT you can do to it (eg. checkbox and radio behave differently, because that's how the HTML spec works).

### AFTER - What is happening after this PR?

Doing the most common things with JS in a form is a lot easier, because Backpack provides:
- a common way to interact with fields, that looks A LOT like our fluent PHP syntax;
- examples for the most common scenarios;

So.. no matter what field you're trying to select or to target, you can do `crud.field('price')` and you'll get back an object with the things you need most of times:
- properties: `name`, `wrapper`, `input`, `value`
- methods for events: `onChange()` -  also aliased `change()`
- methods that do stuff: `hide()`, `show()`, `enable()`, `disable()`, `require()`, `unrequire()`, `check()`, `uncheck()`

With this nifty little library, to show/hide one field depending on another, you'd now just need to:

**Step 1.** Create a script with whatever you want to do (eg. in `public/assets/js/admin/forms/product.js`):

```js
 crud.field('visible').onChange(function(field) {
    if (field.value == 1) {
        crud.field('visible_where').show();
    } else {
        crud.field('visible_where').hide();
    }
 });
```

**Step 2.** Load that script inside the Create/Update form, by doing this in your `setupCreateOperation()` or `setupUpdateOperation()`:

```php
        Widget::add()->type('script')->content('public/assets/js/admin/forms/product.js');
```

Optionally, we _could_ eliminate the second step, using a naming convention. Backpack can check if a JS file exists that follows the naming convention - if yes, it could load it automatically. Not sure it's worth doing, though...

---

This small library has been specifically to cover all scenarios identified as "most common" in here: https://github.com/Laravel-Backpack/CRUD/issues/4158#issuecomment-1038014077 - and it does. You'll see in the examples below... they're all covered. Here's a video of me showing them off too:

https://user-images.githubusercontent.com/1032474/161794212-f873043a-95bb-4d42-a30c-114f7fb30f9f.mp4

### How did you achieve that, in technical terms?

Created a small JS library for it. Just like we have `crud.table` on the ListOperation, now you can work with `crud.field()` and `crud.fields()` inside the Create and Update operations.

### Is it a breaking change?

Non-breaking. But the syntax and functionality might not be 100% final - that's up for debate. 

### How can we test the before & after?

**Step 1.** Use this branch:

```
composer require backpack/crud:"dev-crud-form-javascript-library as 5.9.99"
```

**Step 2.** Inside a CrudController (for example in ProductCrudController in the demo), add the following to your `setupCreateOperation()` or `setupUpdateOperation()` to add the fields for the 10 most common examples: https://gist.github.com/tabacitu/248dd59da9b33debc26cb7496f205bb5#file-productcrudcontroller-php

**Step 3.** Create that `public/assets/js/product-form.js` file, and place the associated JS code for those 10 examples inside it: https://gist.github.com/tabacitu/248dd59da9b33debc26cb7496f205bb5#file-product-form-js

### CODE REVIEW

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
+<script>
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:1]: Why add this in a `blade` file, rather than `js` file? Because I expect it'll suffer changes for a while, and we don't want people to need to publish it, in order to get the updates. When it's stable, we can of course move this to a `js` file.

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     *
+     * Makes it dead-simple for the developer to perform the most common
+     * javascript manipulations, and makes it easy to do custom stuff
+     * too, by exposing the main components (name, wrapper, input).
+     */
+    class CrudField {
+        constructor(fieldName) {
+            this.name = fieldName;
+            this.wrapper = $("[bp-field-name="+ this.name +"]");
+            this.input = $("[bp-field-main-input][name="+ this.name +"]");
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:13]: Notice this. If a field wants to define that its information is not ACTUALLY stored in the first input... but in a different input... it can do that, by just adding `bp-field-main-input` to that input/select/textarea.

I went through the main Backpack fields and it doesn't look like this is needed right now... all the fields that we have right now, if they store the info in a hidden field, that hidden field is the first one. But custom fields and future fields might not respect this convention - for those, we have `bp-field-main-input`.

Of course, I could also be wrong, and we _do_ have fields like that.

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        field: function(fieldName) {
+            return new CrudField(fieldName);
+        },
+        fields: function(fieldNamesArray) {
+            return fieldNamesArray.map(function(fieldName) {
+                return new CrudField(fieldName);
+            });
+        }
+    }
+</script>
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:95]: I'm gonna say this, because if I don't, I know Antonio will 😅  Yes, we can _easily_ eliminate jQuery from this... I know! It was just easier for me to do it in jQuery 😀

Comment (tabacitu) [reply src/resources/views/crud/inc/form_fields_script.blade.php:95]: Also, if we _do_ remove it, then the selectors will NOT be jQuery objects, so you won't be able to do stuff on them with jQuery, you'd have to wrap them... so... yeah... food for thought.

**Comment by tabacitu** [comment]
TODOs to move this forward:
- [ ] get feedback;
- [ ] fix the unrelated problem where widgets are loaded twice (why?!?!?!);
- [ ] test on all field types;
- [ ] try different complex scenarios, to see if our current syntax and API work well for them;
- [ ] triple-check property and method names; once we publish this... we can no longer change them;
- [ ] triple-check closure parameters; once we publish... that's it;
- [ ] decide whether to automatically load a JS file, using a naming convention;
- [ ] write documentation for it (including all 10 examples);
- [ ] (maybe, a little later on) create a PHP syntax, as a separate PR, that would automatically write the necessary JS for the most common scenarios; I still think [the PHP syntax we explained here](https://github.com/Laravel-Backpack/CRUD/issues/4158#issuecomment-1037871469) is a good one, but as single-dimensional arrays: 

```php
CRUD::field([
    'name' => 'extra_cost_value',
    'type' => 'number',
    'showWhen' => [ 'agreed', 'isChecked', 'AND', 'start_date', '>=', 'end_date', 'AND', 'end_date', '<=', date('Y-m-d') ], 
    'hideWhen' => [ 'agreed', 'isUnchecked', 'OR', 'start_date', '<', 'end_date', 'OR', 'end_date', '>', date('Y-m-d') ], 
]);
```

**Comment by margarizaldi** [comment]
Can't wait for this feature release. Please do your magic soon.

**Comment by tabacitu** [comment]
Thank you @margarizaldi ! @pxpm please take a look and give me your take on it. Let's move this forward.

tipsyelves-rainbow-magic-3o84U6421OOWegpQhq

**Comment by rroblik** [comment]
I think I know the answer but... as it is not related to "pro" and this is a basic feature can we hope a v4 backport? 

Thanks

**Comment by tabacitu** [comment]
> I think I know the answer but... as it is not related to "pro" and this is a basic feature can we hope a v4 backport?

Nope... Backpack v4 will only receive security updates in the future, no extra features. See [version support](https://github.com/Laravel-Backpack/CRUD/blob/v5/SECURITY.md#supported-versions) here. We must focus on making v5 the best we can.

**Comment by rroblik** [comment]
Final reason to make me leave BackPack adventure with 🫤 emoji !

**Comment by pxpm** [comment]
> Final reason to make me leave BackPack adventure with 🫤 emoji !

Hey @rroblik sorry to hear that. Like @tabacitu said, we cannot focus on v5 roadmap if we kept pushing stuff into v4, but let me just add that there is nothing preventing you from using this in your v4 project, the code is here, under "Files", so you can just copy it to your project and use it. 

Please don't think I am arguing, I am 100% in favor of finding a better suite for my needs if the current one does not suffice them and I hope you find yours. 

Cheers

**Review by pxpm — CHANGES_REQUESTED** [review]
Hey @tabacitu amazing work here! So I am just giving this a round up review, I am not going to work on fixes unless you tell me to.

So until now I figured out that:
1) It does not work with arrayed name fields like `date_range`. It will throw an error in wrapper because of the `bp-field-name`

2) jquery `date_picker` and `datetime_picker` catch the change event when initializing, but then don't catch the change when you select a date from the picker. (also happens with: `address_algolia` fields)

3) In the `date` html field, if you are manually typing the date, the change event is triggered in every keystoke when you start inputing the year. Same happens with `datetime, time, month, week`

```javascript
crud.field('date').onChange(function(e, value) {
    console.log('date')
    var now = new Date();
    var today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    var selectedDate = Date.parse(value);
    if (selectedDate < today) {
        new Noty({
            type: "error",
            text: "date cant be lower than today"
          }).show();
    }
});
```

4) for hasOne with dot notation the `bp-field-name` should probably be `address.street` so developer can target it with `address.street`, otherwise it needs to know beforehand that we will change the name to `address[street]`. 
- At the moment trying to add `address.street` as a field name in javascript will break the jquery search, changing to passing full strings to the search query makes it work:
```javascript
constructor(fieldName) {
            this.name = fieldName;
            this.wrapper = $('[bp-field-name="'+ this.name +'"]');
            this.input = $('[bp-field-main-input][name="'+ this.name +'"]');
            // if no bp-field-main-input has been declared in the field itself,
            // assume it's the first input in that wrapper, whatever it is
            if (this.input.length == 0) {
                this.input = $('[bp-field-name="'+ this.name +'"] input, [bp-field-name="'+ this.name +'"] textarea, [bp-field-name="'+ this.name +'"] select').first();
            }
            this.value = this.input.val();
        }
```
5) The multiple selects use an hidden input with empty value, not to store the field itself but to submit when nothing is selected. I could not target the selects with `name` or `name[]`. I tried to add:
```php
'wrapper' => [
   'class' => 'form-group col-md-6',
   'bp-field-main-input' => 'name[]'
],
```
No avail here. 

6) can't make any js editor work with onChange, neither by typing or moving focus out. 

7) video field does not catch the event too.

8) `color_picker` jquery plugin will trigger the change event when you drag to chose the color triggering hundreds of times in a short period.

9) `icon_picker` does not work too

10) I haven't tried this with repeatable, but we should consider the "events/hooks" aproach for table as this script has no effect on table field. 

From what I see it's nothing unfixable, we can work this out. 

See you in a few to discuss this!

Cheers

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        change(closure) {
+            this.input.change(function(event) {
+                var fieldWrapper = $(this).parent('[bp-field-wrapper=true]');
+                var fieldName = fieldWrapper.attr('bp-field-name');
+                var fieldType = fieldWrapper.attr('bp-field-type');
+                var fieldValue = $(this).val();
+
+                // console.log('Changed field ' + fieldName + ' (type '+ fieldType + '), value is now ' + fieldValue);
+                closure(event, fieldValue, fieldName, fieldType);
+            }).change();
```
Comment (pxpm) [inline src/resources/views/crud/inc/form_fields_script.blade.php:31]:
```suggestion
            });
```

Any particular reason why this change is needed here ? I've tested and it seems it's not needed at all? What am I missing  @tabacitu ? 🙃

Comment (tabacitu) [reply src/resources/views/crud/inc/form_fields_script.blade.php:31]: I had a reason. But I can't remember it 😅 Maybe the order in which the scripts are run? To make sure if the API gets run _after_ the custom JS, it still works? Anyway... I've tested without it to... seems to work soo... yeah, let's remove it. Good catch @pxpm !

Comment (tabacitu) [reply src/resources/views/crud/inc/form_fields_script.blade.php:31]: Wait! Found it! Pushed a commit that adds back the `.change()`.

Take a look at the "Top 10 Scenarios" tab. Without the `.change()` some of the scenarios won't work. They will 
- have fields visible when the page is loaded, then hide them (flash of fields before they're hidden)
- leave fields visible when they should be hidden

Comment (pxpm) [reply src/resources/views/crud/inc/form_fields_script.blade.php:31]: Hey @tabacitu !

I've been testing this with and without the change and I still think that we should **NOT** call the change if the field didn't change.

The issue about "flashing fields" is just what it is, we can show a "blured form" or a total overlay while form is loading, it's just how it is, html > then > js, it happens with wysiwyg editors too, you can see the textarea before they are turned into editors.

About your second "leave fields visisble" I think you are refering to the `custom_type` in the demo, that should only be visible when value of radio equals custom.

I get what you did here and why you did it, but I think we can't fix that problem the way you did, as that will arise other problems. 

With `.change()` there, we init the html field, and call `change` on it (nothing have changed, still no radio selected, but somewhow it changed).

Try adding validation to a field ? Impossible, if you are validating a field for value empty value, when you have an empty input on the page first load you will have the error triggered.

What I propose to fix this is using another event, like `init` instead of `change`, to separate the initial state from the real change event. 

```javascript
crud.field(monsterFields[i]).onChange(function(e, fieldValue, fieldName, fieldType) {

            if(fieldValue.length <= 0 || typeof fieldValue == "undefined") {
                console.log('gimme value') //add error class, alert user, sirene sound
            }
            console.log('Value for field ' + fieldName + ' (type ' + fieldType +') was changed to '+ fieldValue);
        });
```
Add this on demo `onChange()` test script you will see what I mean.

Please let's think about a solution here, change definitely can't be used for this purpose.

**Comment by maverjk** [comment]
Personally, I really think that this small library is more than helpful. It can improve quite a lot the whole user experience and can help to customize forms in a new way for Backpack. 
I would like that there will be automatically load of JS file (using a naming convention) and I would like also have the possibility to retrieve data from an AJAX request. 

About my last point I'll try to be more explicit: 
I don't know if it's already possible, however let's suppose that I'm going to use a "select2_from_ajax". This field return the whole details of a related entry and it's going to show just some of them (example a "product name"). Well, it would be nice to have the possibility to update others fields value using data returned from my select2 ajax values.

**Comment by margarizaldi** [comment]
> I don't know if it's already possible, however let's suppose that I'm going to use a "select2_from_ajax". This field return the whole details of a related entry and it's going to show just some of them (example a "product name"). Well, it would be nice to have the possibility to update others fields value using data returned from my select2 ajax values.
> 

I also wonder how to achieve this. Usually when we use plain select2, we have full control over the select2 events as well as the data, so setting values of other fields from the response is very handy.

And I think this will be possible if we have select2 event attributes such as:

```
->type('select2_from_ajax')
->onSelect(function () {
    crud.field('price').value = params.data.price
});

```

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            this.input?.addEventListener('input', fieldChanged, false);
+            this.$input.change(fieldChanged);
+            fieldChanged();
+
+            return this;
+        }
+
+        onChange(closure) {
+            this.change(closure);
+        }
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:85]: If we do not `return` in `onChange` then:
- the return of `change` will be a CrudField
- the return of `onChange` will be undefined

In order to make them interchangeable, we need to return the same thing:

```suggestion
        onChange(closure) {
            return this.change(closure);
        }
```

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                console.error(`CrudField error! Could not select INPUT for "${this.name}"`);
+            }
+        }
+
+        get value() {
+            let value = this.input.value;
+
+            // Parse the value if it's a number
+            if (value.length && !isNaN(value)) {
+                value = Number(value);
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:33]: @promatik are you sure this doesn't have unintended consequences? Here's what @pxpm had to say about it on Discord:

> I think I found a quite big inconsistency here:
> - In a text field, if you want to validate the length of the text you would do something like: `field.value.length <= 10`, it works as long as you write atleast one letter, if you happen to use all numbers like: 12345 it will be converted to a real number, and `field.value.length` will return undefined ...

Sounds to me like we have a reason NOT to convert to Number?

Comment (pxpm) [reply src/resources/views/crud/inc/form_fields_script.blade.php:33]: I think there is also a good reason to convert the "real numbers" into numbers, so I proposed that we add a `parseIntegers` or something as a method in `CrudField` so if I am targetting a `number` field I am sure I will use `crud.field('number').change(function(){}).parseIntegers()`

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+            }
+
+            return value;
```
Comment (promatik) [inline src/resources/views/crud/inc/form_fields_script.blade.php:36]:
```suggestion
            return this.input.value;
```

**Review by promatik — CHANGES_REQUESTED** [review]

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+                window.crud.subfieldsCallbacks[this.subfieldHolder] = window.crud.subfieldsCallbacks[this.subfieldHolder] ?? [];
+                if(!window.crud.subfieldsCallbacks[this.subfieldHolder].some( callbacks => callbacks['fieldName'] === this.name )) {
+                    window.crud.subfieldsCallbacks[this.subfieldHolder].push({fieldName: this.name, closure: closure, field: this});
+                }
+                return this;
+            }
+
+            if(['text', 'textarea'].includes(this.input.getAttribute('type'))) {
+                this.input?.addEventListener('input', fieldChanged, false);
+            }
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/inc/form_fields_script.blade.php]: Found a bug that might be related to this. In my tests, inside the SIMPLE tab, only the "text" field now triggers the `change()` event after each key press. The rest of the fields, even tough they are technically INPUT or TEXTAREA... do not, they only trigger it on focusout:

https://user-images.githubusercontent.com/1032474/172439286-58fdaf76-80d1-4304-8740-ef6c46d6d45f.mp4

[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/inc/form_fields_script.blade.php]: @tabacitu can't reproduce. I am on the three main branches with all last commits: https://recordit.co/sYesm7qSnk

[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/inc/form_fields_script.blade.php]: I will post what I posted on discord.

We need to add `?` into this input, or it will raise an error when the field does not exist.

image

Sorry I did the sarcastic comment about the conditional if, it migth seem that I was kidding and not talking serious about using `this.input?` ! My bad. 

It was serious! 😈

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            if(this.isSubfield) {
+                window.crud.subfieldsCallbacks = window.crud.subfieldsCallbacks ?? [];
+                window.crud.subfieldsCallbacks[this.subfieldHolder] = window.crud.subfieldsCallbacks[this.subfieldHolder] ?? [];
+                if(!window.crud.subfieldsCallbacks[this.subfieldHolder].some( callbacks => callbacks['fieldName'] === this.name )) {
+                    window.crud.subfieldsCallbacks[this.subfieldHolder].push({closure: closure, field: this});
+                }
+                return this;
+            }
+
+            if(['INPUT', 'TEXTAREA'].includes(this.input.nodeName)) {
```
Comment (pxpm) [inline src/resources/views/crud/inc/form_fields_script.blade.php:89]:
```suggestion
            if(['INPUT', 'TEXTAREA'].includes(this.input?.nodeName)) {
```

File: src/resources/views/crud/inc/form_fields_script.blade.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/inc/form_fields_script.blade.php]: > We need to add ? into this input, or it will raise an error when the field does not exist.

Ok done!

[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/inc/form_fields_script.blade.php]: About the change not triggered on keyup... I couldn't reproduce now either... 😅 But I have PROOF it was happening. Yeah perhaps some of the merges fixed it. Or I had changes that broke it... 🤷‍♂️ Either way, working now 🎉  Sorry for the false alarm.

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 10]
```diff
@@ -0,0 +1,183 @@
+<script>
+    /**
+     * A front-end representation of a Backpack field, with its main components.
+     *
+     * Makes it dead-simple for the developer to perform the most common
+     * javascript manipulations, and makes it easy to do custom stuff
+     * too, by exposing the main components (name, wrapper, input).
+     */
+    class CrudField {
+        constructor(name) {
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:10]: I've just double-checked the selection logic and it still sounds good to me! Good job here @pxpm ! 👏👏👏

@promatik could you please also give it a look too? Maybe you spot red flags I missed, like you did with looking up through ancestors. Look at `constructor()` but also at `inputWrapper` and `mainInput` please.

File: src/resources/views/crud/fields/upload_multiple.blade.php [hunk 11]
```diff
@@ -163,9 +163,40 @@ function bpFieldInitUploadMultipleElement(element) {
...
+								e.stopImmediatePropagation();
+								return false;
+							});
+							// make the event we just registered, the first to be triggered
+							$._data($deleteButton.get(0), "events").click.reverse();
+						}
+					});
+				});
+
+				element.on('backpack:field.enable', function(e) {
```
Comment (tabacitu) [inline src/resources/views/crud/fields/upload_multiple.blade.php:194]: @pxpm can you please put together the PRs to replace the custom event names, from `backpack:field.something` to `CrudField:something`?

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+        enable(value = true) {
+            this.$input.attr('disabled', !value && 'disabled');
+            this.$input.trigger(`backpack:field.${value ? 'enable' : 'disable'}`);
+            return this;
+        }
+
+        disable(value = true) {
+            return this.enable(!value);
+        }
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:128]: We have `disable()` and `enable()`, but now I'm wondering... In most cases, isn't it `readonly` that people need, not `disabled`? What do you guys think?

Comment (pxpm) [reply src/resources/views/crud/inc/form_fields_script.blade.php:128]: Not sure it's more used than disabled, but definitely it's widely used and we should probably support it.

**Comment by tabacitu** [comment]
**Update:** I've decided NOT to use a convention, to automatically load JS files from a path. Yes, it's one extra step, but:
- that extra step will leave breadcrumbs, which I think is a very good thing for transparency; when you go into that CrudController, you'll see it loads a JS file, and be able to see what it does; otherwise, a new developer that comes into the project can just come into that CrudController and say "WTF" because they don't understand why it behaves in a different way than they expect;
- that extra step is a super-simple one, a one-liner;
- that extra step not only teaches you how to use the CrudField JS API, but also load any other JS or CSS into any other Backpack page;

For those reasons, I've decided that, for now, we won't be automatically loading any JS files onpage. We can change our minds upon feedback, of course. It's easy to add later. But for now... no, let's not do that.

**Another update:** We're SUPER FREAKING CLOSE to launching! 🎉🎉🎉 Everything's working properly, we have docs, we have... EVERYTHING. Running the final checks before we launch 🥳

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            }
+
+            if(['INPUT', 'TEXTAREA'].includes(this.input?.nodeName)) {
+                this.input?.addEventListener('input', fieldChanged, false);
+            }
+            this.$input.change(fieldChanged);
+
+            return this;
+        }
+
```
Comment (tabacitu) [inline src/resources/views/crud/inc/form_fields_script.blade.php:96]: I don't know why we didn't think of this before, but... `.focus()`... I expect people will need to intentionally set the focus on a certain field... after something is done 😔

File: src/resources/views/crud/inc/form_fields_script.blade.php [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        show(value = true) {
+            this.wrapper.toggleClass('d-none', !value);
+            this.$input.trigger(`CrudField:${value ? 'show' : 'hide'}`);
+            return this;
+        }
+
+        hide(value = true) {
+            return this.show(!value);
+        }
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/inc/form_fields_script.blade.php:118]: Sure, It's because the name of that field is not added in the `fields.js` field list.

I have it here, will submit the PR in a few moments.

Cheers

File: src/resources/views/crud/inc/form_fields_script.blade.php (hunk 13 above)
Comment (pxpm) [reply src/resources/views/crud/inc/form_fields_script.blade.php:96]: My only concern about focus is that it needs to be on the current visible elements, it will break if you try to focus on an element in a different tab. 
And then we will be talking about `focus('tab.field')` and adding support to change to the `tab` and focus the `field`. 

Doable ... if it's really something that we need.

**Comment by tabacitu** [comment]
nbc-the-office-oh-my-god-its-happening-huJmPXfeir5JlpPAx0
